File: pg/BO/ssap-partner-portal-bo-frontend/src/types/Project.ts
export interface IProfit {
  monthly: number[]
  quarterly: number[]
  half: number[]
  year: number
}

export interface IGross {
  monthly: number[]
  quarterly: number[]
  half: number[]
  year: number
}

export interface ISalesforceMainCustomer {
  name: string
  email: string
  organizationName: string
}

export interface IProjectListItem {
  id: string
  customerId: string
  salesforceUpdateAt: string
  name: string
  customerName: string
  serviceType: string
  serviceTypeName: string
  createNew: boolean
  mainSalesUserId: string
  mainSalesUserName: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  totalContractTime: number
  mainCustomerUserId: string
  mainCustomerUserName: string
  salesforceMainCustomer: ISalesforceMainCustomer[]
  customerUsers: { key: string }[]
  mainSupporterUserId: string
  mainSupporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  salesforceMainSupporterUserName: string
  supporterUsers: { key: string }[]
  salesforceSupporterUserNames: string[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  isSolverProject: boolean
  updateAt: string
}

export interface IGetProjectsRequest {
  status: string
  fromYearMonth: number
  toYearMonth: number
  customerId: string
  mainSalesUserId: string
  supporterOrganizationId: string
  serviceTypeId: string
  sort: string
  offsetPage: number
  limit: number
}

export interface IGetProjectsResponse {
  offsetPage: number
  total: number
  projects: IProjectListItem[]
}

interface IGetProjectSortDefine {
  NAME_ASC: string
  [key: string]: string
}

export const GET_PROJECTS_REQUEST_SORT: IGetProjectSortDefine = {
  NAME_ASC: 'name:asc',
}
interface ISuggestProjectSortDefine {
  NAME_ASC: string
  [key: string]: string
}

export const SUGGEST_PROJECTS_REQUEST_SORT: ISuggestProjectSortDefine = {
  NAME_ASC: 'name:asc',
}

export interface ISuggestProjectsRequest {
  customerId?: string
  serviceType: string
  sort: string
}

export interface ISuggestProjectResponse {
  id: string
  name: string
  displayName: string
}

export interface ICustomerUser {
  id: string
  name: string
}

export interface ISupporterUser {
  id: string
  name: string
}

export interface IGetProjectByIdResponse {
  id: string
  customerId: string
  salesforceCustomerId: string
  salesforceOpportunityId: string
  salesforceUpdateAt: string
  name: string
  customerName: string
  serviceType: string
  serviceTypeName: string
  createNew: boolean | null
  mainSalesUserId: string
  mainSalesUserName: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit
  totalContractTime?: number
  mainCustomerUserId: string
  mainCustomerUserName: string
  salesforceMainCustomer: ISalesforceMainCustomer
  customerUsers: ICustomerUser[]
  serviceManagerName: string
  mainSupporterUserId: string
  mainSupporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  salesforceMainSupporterUserName: string
  supporterUsers: ISupporterUser[]
  salesforceSupporterUserNames: string[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  dedicatedSurveyUserEmail: string
  dedicatedSurveyUserName: string
  surveyPassword: string
  isSurveyEmailToSalesforceMainCustomer: boolean | null
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export interface ICreateProjectRequest {
  customerId: string
  name: string
  customerName: string
  serviceType: string
  createNew: boolean
  mainSalesUserId: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit[]
  totalContractTime: number
  mainCustomerUserId: string
  salesforceMainCustomer: ISalesforceMainCustomer[]
  customerUsers: { key: string }[]
  mainSupporterUserId: string
  supporterOrganizationId: string
  salesforceMainSupporterUserName: string
  supporterUsers: { key: string }[]
  salesforceSupporterUserNames: string[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  dedicatedSurveyUserEmail: string
  dedicatedSurveyUserName: string
  surveyPassword: string
  isSurveyEmailToSalesforceMainCustomer: boolean | null
}

export interface ICreateProjectResponse {
  id: string
  customerId: string
  salesforceCustomerId: string
  salesforceOpportunityId: string
  salesforceUpdateAt: string
  name: string
  customerName: string
  serviceType: string
  serviceTypeName: string
  createNew: boolean
  mainSalesUserId: string
  mainSalesUserName: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit[]
  totalContractTime: number
  mainCustomerUserId: string
  mainCustomerUserName: string
  salesforceMainCustomer: ISalesforceMainCustomer[]
  customerUsers: { key: string }[]
  mainSupporterUserId: string
  mainSupporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  salesforceMainSupporterUserName: string
  supporterUsers: { key: string }[]
  salesforceSupporterUserNames: string[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  dedicatedSurveyUserEmail: string
  dedicatedSurveyUserName: string
  surveyPassword: string
  isSurveyEmailToSalesforceMainCustomer: boolean | null
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export interface IUpdateProjectRequestParam {
  id: string
  version: number
}

export interface IUpdateProjectRequest {
  customerId: string
  name: string
  customerName: string
  serviceType: string
  createNew: boolean
  mainSalesUserId: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit[]
  totalContractTime: number
  mainCustomerUserId: string
  salesforceMainCustomer: ISalesforceMainCustomer[]
  customerUsers: { key: string }[]
  mainSupporterUserId: string
  supporterOrganizationId: string
  supporterUsers: { key: string }[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  dedicatedSurveyUserEmail: string
  dedicatedSurveyUserName: string
  surveyPassword: string
  isSurveyEmailToSalesforceMainCustomer: boolean | null
}

export interface IUpdateProjectResponse {
  id: string
  customerId: string
  salesforceCustomerId: string
  salesforceOpportunityId: string
  salesforceUpdateAt: string
  name: string
  customerName: string
  serviceType: string
  serviceTypeName: string
  createNew: boolean
  mainSalesUserId: string
  mainSalesUserName: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit[]
  totalContractTime: number
  mainCustomerUserId: string
  mainCustomerUserName: string
  salesforceMainCustomer: ISalesforceMainCustomer[]
  dedicatedSurveyUserEmail: string
  dedicatedSurveyUserName: string
  surveyPassword: string
  isSurveyEmailToSalesforceMainCustomer: boolean | null
  customerUsers: { key: string }[]
  mainSupporterUserId: string
  mainSupporterUserName: string
  supporterOrganizationId: string
  supporterOrganizationName: string
  salesforceMainSupporterUserName: string
  supporterUsers: { key: string }[]
  salesforceSupporterUserNames: string[]
  isCountManHour: boolean
  isKarteRemind: boolean
  isMasterKarteRemind: boolean
  contractType: string
  isSecret: boolean
  createId: string
  createUserName: string
  createAt: string
  updateId: string
  updateUserName: string
  updateAt: string
  version: number
}

export const ENUM_CREATE_OR_UPDATE_PROJECT_RESPONSE_ERROR = {
  DISABLED_MAINSALES: 'disabled mainSales is set',
  DISABLED_MAINCUSTOMER: 'disabled mainCustomer is set',
  DISABLED_CUSTOMER: 'disabled customer is set',
  DISABLED_MAINSUPPORTER: 'disabled mainSupporter is set',
  DISABLED_SUPPORTER: 'disabled supporter is set',
} as const

export interface IDeleteProjectRequestParam {
  id: string
  version: number
}

export interface IDeleteProjectResponse {
  message: string
}

export const ENUM_IMPORT_PROJECTS_MODE = {
  CHECK: 'check',
  EXECUTE: 'execute',
} as const

export type importProjectsMode =
  typeof ENUM_IMPORT_PROJECTS_MODE[keyof typeof ENUM_IMPORT_PROJECTS_MODE]

export const ENUM_IMPORT_PROJECTS_RESULT = {
  OK: 'ok',
  NG: 'ng',
  DONE: 'done',
  ERROR: 'error',
} as const

export type importProjectsResult =
  typeof ENUM_IMPORT_PROJECTS_RESULT[keyof typeof ENUM_IMPORT_PROJECTS_RESULT]

interface IImportedProject {
  salesforceOpportunityId: string
  customerId: string
  salesforceUpdateAt: string
  name: string
  customerName: string
  serviceType: string
  salesforceUsePackage: boolean
  createNew: boolean
  salesforceViaPr: boolean
  salesforceMainCustomer: ISalesforceMainCustomer[]
  mainSalesUserName: string
  phase: string
  customerSuccess: string
  supportDateFrom: string
  supportDateTo: string
  profit: IProfit[]
  gross: IGross[]
  salesforceMainSupporterUserName: string
  supporterUsers: string[]
  totalContractTime: number
  errorMessage: string
}

export class ImportedProject implements IImportedProject {
  salesforceOpportunityId = ''
  customerId = ''
  salesforceUpdateAt = ''
  name = ''
  customerName = ''
  serviceType = ''
  salesforceUsePackage = true
  createNew = true
  salesforceViaPr = true
  salesforceMainCustomer = []
  mainSalesUserName = ''
  phase = ''
  customerSuccess = ''
  supportDateFrom = ''
  supportDateTo = ''
  profit = []
  gross = []
  salesforceMainSupporterUserName = ''
  supporterUsers = []
  totalContractTime = 0
  errorMessage = ''
}

export interface IImportProjectsRequest {
  file: string
  mode: importProjectsMode
}

export interface IImportProjectsResponse {
  mode: importProjectsMode
  result: importProjectsResult
  projects: ImportedProject[]
}

export interface IMember {
  supporterName: string
  role: string
  isConfirm: boolean
}

export interface IOrganization {
  supporterOrganizationName: string
  members: IMember[]
}

export interface IGetMonthlyProjectsRequest {
  year: number
  month: number
}

export interface IMonthlyProjectHeader {
  supporterOrganizationId: string
  supporterOrganizationName: string
}
export interface IGetMonthlyProjectResponse {
  serviceTypeName: string
  name: string
  contractType: string
  mainSalesUserName: string
  Organizations: IOrganization[]
  supportDateFrom: string
  supportDateTo: string
  totalContractTime: number
  thisMonthContractTime: number
  totalProfit: number
  thisMonthProfit: number
}

export interface IGetMonthlyProjectsResponse {
  details: IGetMonthlyProjectResponse[]
  header: IMonthlyProjectHeader[]
}

File: BO/ssap-partner-portal-bo-backend/app/models/project.py
from datetime import datetime

from fastapi import HTTPException, status
from pynamodb.attributes import (
    BooleanAttribute,
    ListAttribute,
    MapAttribute,
    NumberAttribute,
    UnicodeAttribute,
    UnicodeSetAttribute,
    VersionAttribute,
)
from pynamodb.exceptions import DoesNotExist
from pynamodb.indexes import AllProjection, GlobalSecondaryIndex
from pynamodb.models import Model

from app.core.config import get_app_settings
from app.models.custom_attribute.jst_datetime import JSTDateTimeAttribute
from app.resources.const import DataType


class CustomerIdNameIndex(GlobalSecondaryIndex):
    """GSI：顧客ID案件名検索"""

    class Meta:
        index_name = "customer_id-name-index"
        projection = AllProjection()

    customer_id = UnicodeAttribute(hash_key=True)
    name = UnicodeAttribute(range_key=True)


class DataTypeNameIndex(GlobalSecondaryIndex):
    """GSI：データ区分案件名検索"""

    class Meta:
        index_name = "data_type-name-index"
        projection = AllProjection()

    data_type = UnicodeAttribute(hash_key=True)
    name = UnicodeAttribute(range_key=True)


class DataTypeSupportDateToIndex(GlobalSecondaryIndex):
    """GSI：支援終了日検索"""

    class Meta:
        index_name = "data_type-support_date_to-index"
        projection = AllProjection()

    data_type = UnicodeAttribute(hash_key=True)
    support_date_to = UnicodeAttribute(range_key=True)


class GrossProfitAttribute(MapAttribute):
    """案件TBL Attribute定義"""

    monthly = ListAttribute(null=True)
    quarterly = ListAttribute(null=True)
    half = ListAttribute(null=True)
    year = NumberAttribute(null=True)


class SalesforceMainCustomerAttribute(MapAttribute):
    """案件TBL Attribute定義"""

    name = UnicodeAttribute(null=True)
    email = UnicodeAttribute(null=True)
    organization_name = UnicodeAttribute(null=True)
    job = UnicodeAttribute(null=True)


class UpdateAttributesSubAttribute(MapAttribute):
    """案件TBL Attribute定義"""

    attribute = UnicodeAttribute(null=True)
    value = UnicodeAttribute(null=True)


class UpdateHistoryAttribute(MapAttribute):
    """案件TBL Attribute定義"""

    update_id = UnicodeAttribute(null=True)
    update_attributes = ListAttribute(of=UpdateAttributesSubAttribute, null=True)


class ProjectModel(Model):
    """案件TBLモデル定義"""

    class Meta:
        table_name = get_app_settings().table_prefix + "Project"
        region = get_app_settings().aws_region

    id = UnicodeAttribute(hash_key=True, null=False)
    data_type = UnicodeAttribute(range_key=True, null=False)
    customer_id = UnicodeAttribute(null=False)
    salesforce_customer_id = UnicodeAttribute(null=True)
    salesforce_opportunity_id = UnicodeAttribute(null=True)
    salesforce_update_at = JSTDateTimeAttribute(null=True)
    name = UnicodeAttribute(null=False)
    customer_name = UnicodeAttribute(null=True)
    service_type = UnicodeAttribute(null=True)
    create_new = BooleanAttribute(null=True)
    continued = BooleanAttribute(null=True)
    main_sales_user_id = UnicodeAttribute(null=False)
    contract_date = UnicodeAttribute(null=True)
    phase = UnicodeAttribute(null=True)
    customer_success = UnicodeAttribute(null=True)
    support_date_from = UnicodeAttribute(null=False)
    support_date_to = UnicodeAttribute(null=False)
    profit = GrossProfitAttribute(null=True)
    gross = GrossProfitAttribute(null=True)
    total_contract_time = NumberAttribute(null=False)
    main_customer_user_id = UnicodeAttribute(null=True)
    salesforce_main_customer = SalesforceMainCustomerAttribute(null=True)
    customer_user_ids = UnicodeSetAttribute(null=True)
    service_manager_name = UnicodeAttribute(null=True)
    main_supporter_user_id = UnicodeAttribute(null=True)
    supporter_organization_id = UnicodeAttribute(null=True)
    salesforce_main_supporter_user_name = UnicodeAttribute(null=True)
    supporter_user_ids = UnicodeSetAttribute(null=True)
    salesforce_supporter_user_names = UnicodeSetAttribute(null=True)
    is_count_man_hour = BooleanAttribute(null=False, default=True)
    is_karte_remind = BooleanAttribute(null=False, default=True)
    is_master_karte_remind = BooleanAttribute(null=False, default=True)
    contract_type = UnicodeAttribute(null=True)
    is_secret = BooleanAttribute(null=False, default=False)
    is_solver_project = BooleanAttribute(default=False)
    salesforce_use_package = BooleanAttribute(null=True)
    salesforce_via_pr = BooleanAttribute(null=True)
    update_history = UpdateHistoryAttribute(null=True)
    survey_group_id = UnicodeAttribute(null=True)
    is_man_hour_input = BooleanAttribute(null=True, default=False)
    dedicated_survey_user_name = UnicodeAttribute(null=True)
    dedicated_survey_user_email = UnicodeAttribute(null=True)
    survey_password = UnicodeAttribute(null=True)
    is_survey_email_to_salesforce_main_customer = BooleanAttribute(null=True)
    this_month_type = UnicodeAttribute(null=True)
    no_send_reason = UnicodeAttribute(null=True)

    create_id = UnicodeAttribute(null=True)
    create_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    update_id = UnicodeAttribute(null=True)
    update_at = JSTDateTimeAttribute(null=False, default=datetime.now())
    version = VersionAttribute()

    customer_id_name_index = CustomerIdNameIndex()
    data_type_support_date_to_index = DataTypeSupportDateToIndex()
    data_type_name_index = DataTypeNameIndex()

    @staticmethod
    def get_project(project_id: str):
        try:
            return ProjectModel.get(hash_key=project_id, range_key=DataType.PROJECT)
        except DoesNotExist:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Project not found."
            )
